index.js

```diff
@@ -2140,18 +2140,16 @@
     });
 
     if (Array.isArray(foldable)) {
-      // If the original foldable was an array we don't need to convert the
-      // array back to a foldable.
       for (var idx = 0; idx < rs.length; idx += 1) rs[idx] = rs[idx].x;
       return rs;
-    } else {
-      var result = empty(foldable.constructor);
-      var of = foldable.constructor['fantasy-land/of'];
-      for (idx = 0; idx < rs.length; idx += 1) {
-        result = result['fantasy-land/concat'](of(rs[idx].x));
-      }
-      return result;
     }
+
+    var F = foldable.constructor;
+    var result = empty(F);
+    for (idx = 0; idx < rs.length; idx += 1) {
+      result = concat(result, of(F, rs[idx].x));
+    }
+    return result;
```

Comment:
This does seem to affect performance. On my computer, the Foldable benchmark went from 308 Hz to 233 Hz. That is, the previous is about 33% faster.

Reply:
I'm concerned about correctness. Representatives of built-in types do not have `fantasy-land/`-prefixed members, so replacing `concat(x, y)` with `x['fantasy-land/concat'](y)` is not necessarily valid. Can we be sure that `result` will have a `fantasy-land/concat` member and that `F` will have a `fantasy-land/of` member?

> `fantasy-land/concat` implementations are provided for the following built-in types: String, Array, and Object.

> `fantasy-land/of` implementations are provided for the following built-in types: Array and Function.

Reply:
Are any of the build-in types, except for Array, both a foldable and a monoid?

Reply:
As a `StrMap`, `Object` also implements `Foldable` and `Monoid`.

Reply:
@gabejohnson Sorry, I forgot to mention that it has to be an Applicative as well.

Reply:
Ahhh...I missed the `of`. Yeah, `Array` is the only one that fits the bill AFAIK.

Reply:
If we reinstate the `['fantasy-land/concat']` and `['fantasy-land/of']` optimizations, let's add a comment explaining why these optimizations are safe.